### PR TITLE
TEST/IODEMO: Add server timeout on handling connect requests.

### PR DESCRIPTION
**Pull request by alinask**
If the server could not respond to a conn_request in a
certain peroid of time, reject the connect request.
The period of time is the value of _connect_timeout.

### CODE REVIEW

**Comment by alinask** [comment]
@yosefe Can you please review?

File: test/apps/iodemo/ucx_wrapper.cc [hunk 1]
```diff
@@ -215,7 +216,13 @@ void UcxContext::connect_callback(ucp_conn_request_h conn_req, void *arg)
                 << ucs_status_string(status) << ")";
     }
 
-    self->_conn_requests.push_back(conn_req);
+    struct timeval tv_current;
+    gettimeofday(&tv_current, NULL);
+
+    conn_request.conn_request = conn_req;
+    conn_request.arrival_time = tv_current;
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc]: gettimeofday(&conn_request.arrival_time, NULL)

File: test/apps/iodemo/ucx_wrapper.cc [hunk 2]
```diff
@@ -271,13 +278,25 @@ double UcxContext::connect_timeout() const
...
-        UcxConnection *conn = new UcxConnection(*this, get_next_conn_id());
-        if (conn->accept(_conn_requests.front())) {
+        UcxConnection *conn     = new UcxConnection(*this, get_next_conn_id());
+        conn_req_t conn_request = _conn_requests.front();
+
+        struct timeval tv_current, elapsed;
+        gettimeofday(&tv_current, NULL);
+
+        timersub(&tv_current, &conn_request.arrival_time, &elapsed);
+        if (elapsed.tv_sec + (elapsed.tv_usec * 1e-6) > _connect_timeout) {
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc]: unite common code with wait_completion()

File: test/apps/iodemo/ucx_wrapper.cc [hunk 3]
```diff
@@ -271,13 +278,25 @@ double UcxContext::connect_timeout() const
...
+            UCX_LOG << "reject connection request " << conn_request.conn_request
+                    << " since server's timeout (" << _connect_timeout
+                    << ") elapsed";
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc]: pls print time units

File: test/apps/iodemo/ucx_wrapper.cc [hunk 4]
```diff
@@ -388,8 +407,8 @@ void UcxContext::handle_connection_error(UcxConnection *conn)
 void UcxContext::destroy_connections()
 {
     while (!_conn_requests.empty()) {
-        UCX_LOG << "reject connection request " << _conn_requests.front();
-        ucp_listener_reject(_listener, _conn_requests.front());
+        UCX_LOG << "reject connection request " << _conn_requests.front().conn_request;
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc]: pls add local var for _conn_requests.front().conn_request

**Comment by alinask** [comment]
@yosefe updated. please take a look.

File: test/apps/iodemo/ucx_wrapper.cc [hunk 5]
```diff
@@ -268,16 +272,40 @@ double UcxContext::connect_timeout() const
     return _connect_timeout;
 }
 
+int UcxContext::is_timeout_elapsed(struct timeval *tv_later,
+                                   struct timeval *tv_prior, double timeout)
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc]: const*

File: test/apps/iodemo/ucx_wrapper.cc [hunk 6]
```diff
@@ -268,16 +272,40 @@ double UcxContext::connect_timeout() const
...
+{
+    struct timeval elapsed;
+
+    timersub(tv_later, tv_prior, &elapsed);
+    if (elapsed.tv_sec + (elapsed.tv_usec * 1e-6) > timeout) {
+        return 1;
+    }
+    return 0;
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc]: return (elapsed.tv_sec + (elapsed.tv_usec * 1e-6) > timeout)

File: test/apps/iodemo/ucx_wrapper.cc [hunk 7]
```diff
@@ -268,16 +272,40 @@ double UcxContext::connect_timeout() const
...
 void UcxContext::progress_conn_requests()
 {
     while (!_conn_requests.empty()) {
-        UcxConnection *conn = new UcxConnection(*this, get_next_conn_id());
-        if (conn->accept(_conn_requests.front())) {
+        UcxConnection *conn     = new UcxConnection(*this, get_next_conn_id());
+        conn_req_t conn_request = _conn_requests.front();
+
+        struct timeval tv_current;
+        gettimeofday(&tv_current, NULL);
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc]: put this into is_timeout_elapsed as well

Comment (alinask) [reply test/apps/iodemo/ucx_wrapper.cc]: do we always want to compare with current time? why not leave it as an arg?

File: test/apps/iodemo/ucx_wrapper.cc (hunk 5 above)
Comment (alinask) [reply test/apps/iodemo/ucx_wrapper.cc]: you mean?
>int is_timeout_elapsed(struct timeval const *tv_later, struct timeval const *tv_prior,  double timeout);

File: test/apps/iodemo/ucx_wrapper.cc (hunk 7 above)
Comment (yosefe) [reply test/apps/iodemo/ucx_wrapper.cc]: yes, to unite common code

File: test/apps/iodemo/ucx_wrapper.cc (hunk 5 above)
Comment (yosefe) [reply test/apps/iodemo/ucx_wrapper.cc]: yes (except not passing current)

**Comment by alinask** [comment]
@yosefe please review again.

**Review by yosefe — APPROVED** [review]

**Pull request merged**